In [1]:
from hpsklearn import HyperoptEstimator, any_regressor, any_classifier, any_preprocessing
from sklearn.model_selection import train_test_split
from hyperopt import tpe
import pandas as pd
import numpy as np

WARN: OMP_NUM_THREADS=None =>
... If you are using openblas if you are using openblas set OMP_NUM_THREADS=1 or risk subprocess calls hanging indefinitely


In [2]:
dtypes = {'age': np.float64, 'Medu': np.float64, 'Fedu': np.float64, 'traveltime': np.float64, 'studytime': np.float64, 'failures': np.float64, 'famrel': np.float64, 'freetime': np.float64, 'goout': np.float64, 'Dalc': np.float64, 'Walc': np.float64, 'health': np.float64, 'absences': np.float64, 'G1': np.float64, 'G2': np.float64, 'G3': np.float64, 'improved': np.float64, 'school_MS': np.float64, 'sex_M': np.float64, 'address_U': np.float64, 'famsize_LE3': np.float64, 'Pstatus_T': np.float64, 'Mjob_health': np.float64, 'Mjob_other': np.float64, 'Mjob_services': np.float64, 'Mjob_teacher': np.float64, 'Fjob_health': np.float64, 'Fjob_other': np.float64, 'Fjob_services': np.float64, 'Fjob_teacher': np.float64, 'reason_home': np.float64, 'reason_other': np.float64, 'reason_reputation': np.float64, 'guardian_mother': np.float64, 'guardian_other': np.float64, 'schoolsup_yes': np.float64, 'famsup_yes': np.float64, 'paid_yes': np.float64, 'activities_yes': np.float64, 'nursery_yes': np.float64, 'higher_yes': np.float64, 'internet_yes': np.float64, 'romantic_yes': np.float64}
data = pd.read_csv('../data_processing/cleaned-data.csv', index_col=0, dtype=dtypes)

y = data.G3.values
y.reshape(-1,1)

X = data.drop(['G1', 'G2', 'G3'], axis=1).values

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

In [4]:
import warnings
warnings.filterwarnings("ignore", category=DeprecationWarning)

estim = HyperoptEstimator(classifier=any_classifier('my_clf'),
                          preprocessing=any_preprocessing('my_pre'),
                          algo=tpe.suggest,
                          max_evals=100,
                          trial_timeout=120)

# Search the hyperparameter space based on the data

estim.fit(X_train, y_train)

# Show the results

print(estim.score(X_test, y_test))

print( estim.best_model() )

0.16666666666666666
{'learner': AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=0.09875123752699139, n_estimators=43,
          random_state=0), 'preprocs': (Normalizer(copy=True, norm='l1'),), 'ex_preprocs': ()}


In [5]:
estim_reg = HyperoptEstimator(regressor=any_regressor('my_clf'),
                          preprocessing=any_preprocessing('my_pre'),
                          algo=tpe.suggest,
                          max_evals=100,
                          trial_timeout=120)

# Search the hyperparameter space based on the data

estim.fit(X_train, y_train)

# Show the results

print(estim.score(X_test, y_test))

print( estim.best_model() )

0.12037037037037036
{'learner': AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=0.0017147934290458798, n_estimators=148,
          random_state=0), 'preprocs': (), 'ex_preprocs': ()}


# Regression First Run
0.12037037037037036
{'learner': GradientBoostingClassifier(criterion='friedman_mse', init=None,
              learning_rate=0.019416675115989265, loss='deviance',
              max_depth=None, max_features=0.603513235739145,
              max_leaf_nodes=None, min_impurity_decrease=0.0,
              min_impurity_split=None, min_samples_leaf=23,
              min_samples_split=2, min_weight_fraction_leaf=0.0,
              n_estimators=27, n_iter_no_change=None, presort='auto',
              random_state=1, subsample=0.7471292832416112, tol=0.0001,
              validation_fraction=0.1, verbose=0, warm_start=False), 'preprocs': (MinMaxScaler(copy=True, feature_range=(0.0, 1.0)),), 'ex_preprocs': ()}
              
              
# Classification First Run
0.1111111111111111
{'learner': KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='euclidean',
           metric_params=None, n_jobs=1, n_neighbors=41, p=2,
           weights='distance'), 'preprocs': (PCA(copy=True, iterated_power='auto', n_components=40, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False),), 'ex_preprocs': ()}

# Regression Second Run
0.12962962962962962
{'learner': AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=0.01093619047791917, n_estimators=11,
          random_state=0), 'preprocs': (PCA(copy=True, iterated_power='auto', n_components=40, random_state=None,
  svd_solver='auto', tol=0.0, whiten=True),), 'ex_preprocs': ()}

# Classification Second Run
0.1388888888888889
{'learner': RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features='log2', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=13, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=466, n_jobs=1,
            oob_score=False, random_state=2, verbose=False,
            warm_start=False), 'preprocs': (Normalizer(copy=True, norm='l2'),), 'ex_preprocs': ()}

# Regression Third Run
0.16666666666666666
{'learner': AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=0.09875123752699139, n_estimators=43,
          random_state=0), 'preprocs': (Normalizer(copy=True, norm='l1'),), 'ex_preprocs': ()}

# Classification Third Run
0.12037037037037036
{'learner': AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None,
          learning_rate=0.0017147934290458798, n_estimators=148,
          random_state=0), 'preprocs': (), 'ex_preprocs': ()}